### PR TITLE
networkd: Add DHCP  support for blacklisting servers

### LINKED ISSUE #6260: systemd-netword - [DHCP] Request support for blacklisting servers
### Submission type

 - Request for enhancement (RFE)

### systemd version the issue has been seen with

> 233

### Used distribution

> NixOS

### Details

dhcpcd is allowed to ignore all packets from a certain DHCP server address via the `blacklist` option ([man 5 dhcpcd.conf](https://www.daemon-systems.org/man/dhcpcd.conf.5.html)). I would like to see this feature on systemd-networkd, too.

### PR DESCRIPTION
**Pull request by ssahani**
closes #6260

### CODE REVIEW

File: src/network/networkd-network.c [hunk 1]
```diff
@@ -1554,6 +1554,70 @@ int config_parse_section_route_table(
...
+                                   "DHCP black listed ip address is invalid, ignoring assignment: %s", n);
+                        continue;
+                }
+
+                r = set_ensure_allocated(&network->dhcp_black_listed_ip, NULL);
+                if (r < 0)
+                        return log_oom();
+
+                r = set_put(network->dhcp_black_listed_ip, UINT32_TO_PTR(ip.in.s_addr));
+                if (r < 0) {
```
Comment (yuwata) [inline src/network/networkd-network.c]: Maybe, `r == -EEXIST` ?

File: src/network/networkd-network.c [hunk 2]
```diff
@@ -1554,6 +1554,70 @@ int config_parse_section_route_table(
...
+                        log_syntax(unit, LOG_WARNING, filename, line, r,
+                                   "DHCP black listed ip address is duplicated, ignoring assignment: %s", n);
+                }
```
Comment (yuwata) [inline src/network/networkd-network.c:1679]: then missing `continue;`?

File: src/network/networkd-dhcp4.c [hunk 3]
```diff
@@ -511,7 +511,45 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
...
+        struct in_addr *i;
+        int r;
+
+        assert(link);
+        assert(link->network);
+        assert(client);
+
+        if (set_size(link->network->dhcp_black_listed_ip) <= 0) {
+                log_link_debug(link, "Blacklist DHCP server DB is empty: %m");
+                return -1;
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: Please make the function return boolean or negative errno. So, this should be `false`.

File: src/network/networkd-dhcp4.c [hunk 4]
```diff
@@ -511,7 +511,45 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
 
         return 0;
 }
-static void dhcp4_handler(sd_dhcp_client *client, int event, void *userdata) {
+
+static int dhcp_server_in_black_listed_ip(Link *link, sd_dhcp_client *client) {
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: Not a matter but I prefer the naming `dhcp_server_is_black_listed()`.

File: src/network/networkd-dhcp4.c [hunk 5]
```diff
@@ -511,7 +511,45 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
...
+        }
+
+        r = sd_dhcp_client_get_lease(client, &lease);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "Failed to get DHCP lease: %m");
+
+        r = sd_dhcp_lease_get_server_identifier(lease, &addr);
+        if (r < 0) {
+                log_link_debug(link, "Failed to get DHCP server ip address: %m");
+                return r;
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: `return log_link_debug_errno();`

File: src/network/networkd-dhcp4.c [hunk 6]
```diff
@@ -511,7 +511,45 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
...
+        }
+
+        i = set_get(link->network->dhcp_black_listed_ip, UINT32_TO_PTR(le32toh(addr.s_addr)));
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: It seems that this and the above `set_size()` can be combined with `set_contains()`.

File: src/network/networkd-dhcp4.c [hunk 7]
```diff
@@ -511,7 +511,45 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
...
+        if (i) {
+                log_struct(LOG_DEBUG,
+                           LOG_LINK_INTERFACE(link),
+                           LOG_LINK_MESSAGE(link, "DHCPv4 ip '%u.%u.%u.%u' found in black listed ip addresses, ignoring offer",
+                                            ADDRESS_FMT_VAL(addr)));
+                return 1;
+        }
+
+        return -ENOENT;
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: and then simply returns the value provided by `set_contains()`.

File: src/network/networkd-dhcp4.c [hunk 8]
```diff
@@ -520,30 +558,30 @@ static void dhcp4_handler(sd_dhcp_client *client, int event, void *userdata) {
...
+                return -1;
 
         switch (event) {
                 case SD_DHCP_CLIENT_EVENT_EXPIRED:
                 case SD_DHCP_CLIENT_EVENT_STOP:
                 case SD_DHCP_CLIENT_EVENT_IP_CHANGE:
                         if (link->network->dhcp_critical) {
                                 log_link_error(link, "DHCPv4 connection considered system critical, ignoring request to reconfigure it.");
-                                return;
+                                return r;
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: `r` is not updated. So, I prefer `return 0;`.

File: src/network/networkd-dhcp4.c [hunk 9]
```diff
@@ -520,30 +558,30 @@ static void dhcp4_handler(sd_dhcp_client *client, int event, void *userdata) {
         assert(link->manager);
 
         if (IN_SET(link->state, LINK_STATE_FAILED, LINK_STATE_LINGER))
-                return;
+                return -1;
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: `-1` is not good. Maybe `0` is enough?

File: src/network/networkd-dhcp4.c [hunk 10]
```diff
@@ -552,16 +590,21 @@ static void dhcp4_handler(sd_dhcp_client *client, int event, void *userdata) {
...
                         if (r < 0) {
                                 link_enter_failed(link);
-                                return;
+                                return r;
                         }
                         break;
+                case SD_DHCP_CLIENT_EVENT_SELECTING:
+                        r = dhcp_server_in_black_listed_ip(link, client);
+                        if (r >= 0)
+                                return -1;
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: Please do not use `-1`.
```
if (r < 0)
        return r;
if (r == 0)
        return -ENOMSG;
```

File: src/network/networkd-dhcp4.c (hunk 4 above)
Comment (ssahani) [reply src/network/networkd-dhcp4.c]: sounds better

File: src/network/networkd-dhcp4.c (hunk 10 above)
Comment (ssahani) [reply src/network/networkd-dhcp4.c]: since it's bool now return ```-ENOMSG``` is enough.

**Comment by ssahani** [comment]
updated thanks for review.

**Comment by yuwata** [comment]
(I guess this needs to be rebased after #12429 will be merged. Hence, I added 'needs-rebase' tag.)

**Comment by ssahani** [comment]
rebased PTAL @yuwata .

**Comment by ssahani** [comment]
@yuwata could please take a look. I will add the DHCP Release feature on top of it's after merging thanks.

**Comment by yuwata** [comment]
Sorry for the delay.

The functionality may be implemented in `sd-dhcp-client.c`, that is, `sd_dhcp_client_set_blacklist_server()` or something. WDYT?

**Comment by ssahani** [comment]
Well let's just utilize the core library to build features top of it. Also feature link whitelisting also is there means it will process the response from those pool of server IP addresses and when this is configured blacklisting is ignored.

File: src/network/networkd-dhcp4.c [hunk 11]
```diff
@@ -512,7 +512,39 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
...
+        int r;
+
+        assert(link);
+        assert(link->network);
+        assert(client);
+
+        r = sd_dhcp_client_get_lease(client, &lease);
+        if (r < 0) {
+                log_link_warning_errno(link, r, "Failed to get DHCP lease: %m");
+                return false;
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: In this case, please return the negative errno. That is,
```
return log_link_warning_errno(link, r, "...");
```

File: src/network/networkd-dhcp4.c [hunk 12]
```diff
@@ -512,7 +512,39 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
...
+        }
+
+        r = sd_dhcp_lease_get_server_identifier(lease, &addr);
+        if (r < 0) {
+                log_link_debug_errno(link, r, "Failed to get DHCP server ip address: %m");
+                return false;
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: Same here.

File: src/network/networkd-dhcp4.c [hunk 13]
```diff
@@ -533,8 +565,8 @@ static void dhcp4_handler(sd_dhcp_client *client, int event, void *userdata) {
 
                                 r = sd_ipv4ll_start(link->ipv4ll);
                                 if (r < 0) {
-                                        log_link_warning_errno(link, r, "Could not acquire IPv4 link-local address: %m");
-                                        return;
+                                        log_link_warning_errno(link, r,  "Could not acquire IPv4 link-local address: %m");
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: Please do not add unnecessary white space.

File: src/network/networkd-dhcp4.c [hunk 14]
```diff
@@ -533,8 +565,8 @@ static void dhcp4_handler(sd_dhcp_client *client, int event, void *userdata) {
...
+                                        return r;
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: `return log_link_warning_errno(link, r, "...");`

File: src/network/networkd-network.c [hunk 15]
```diff
@@ -1620,6 +1620,71 @@ int config_parse_dhcp_max_attempts(
...
+        if (isempty(rvalue)) {
+                network->dhcp_black_listed_ip = set_free(network->dhcp_black_listed_ip);
+                return 0;
+        }
+
+        for (;;) {
+                _cleanup_free_ char *n = NULL;
+                union in_addr_union ip;
+
+                r = extract_first_word(&rvalue, &n, NULL, 0);
```
Comment (yuwata) [inline src/network/networkd-network.c]: Please do not update `rvalue`, as this is used in the log.
```
for (p = rvalue;;) {
    ...
    r = extract_first_word(&p, &n, NULL, 0);
```

File: src/network/networkd-network.c [hunk 16]
```diff
@@ -1620,6 +1620,71 @@ int config_parse_dhcp_max_attempts(
...
+                if (r < 0) {
+                        log_syntax(unit, LOG_ERR, filename, line, r,
+                                   "Failed to parse DHCP black listed ip  address, ignoring assignment: %s",
```
Comment (yuwata) [inline src/network/networkd-network.c]: Doubled white space between `ip` and `address`.

File: src/network/networkd-network.c [hunk 17]
```diff
@@ -1620,6 +1620,71 @@ int config_parse_dhcp_max_attempts(
...
+                        return log_oom();
+
+                r = set_put(network->dhcp_black_listed_ip, UINT32_TO_PTR(ip.in.s_addr));
+                if (r == -EEXIST) {
+                        log_syntax(unit, LOG_WARNING, filename, line, r,
+                                   "DHCP black listed ip address is duplicated, ignoring assignment: %s", n);
+                        continue;
+                }
+
+                if (r < 0)
```
Comment (yuwata) [inline src/network/networkd-network.c:1680]: Please drop the blank line.

File: src/network/networkd-dhcp4.c [hunk 18]
```diff
@@ -512,7 +512,39 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
...
+                return false;
+        }
+
+        r = sd_dhcp_lease_get_server_identifier(lease, &addr);
+        if (r < 0) {
+                log_link_debug_errno(link, r, "Failed to get DHCP server ip address: %m");
+                return false;
+        }
+
+        if (set_contains(link->network->dhcp_black_listed_ip, UINT32_TO_PTR(le32toh(addr.s_addr)))) {
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: Is `le32toh()` necessary?

**Comment by ssahani** [comment]
updated . PTAL.  Thanks @yuwata .

File: src/network/networkd-dhcp4.c [hunk 19]
```diff
@@ -512,7 +512,35 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
         return 0;
 }
 
-static void dhcp4_handler(sd_dhcp_client *client, int event, void *userdata) {
+static bool dhcp_server_is_black_listed(Link *link, sd_dhcp_client *client) {
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: The type of return value should be `int`.

File: src/network/networkd-dhcp4.c [hunk 20]
```diff
@@ -568,16 +594,20 @@ static void dhcp4_handler(sd_dhcp_client *client, int event, void *userdata) {
...
                 case SD_DHCP_CLIENT_EVENT_IP_ACQUIRE:
                         r = dhcp_lease_acquired(client, link);
                         if (r < 0) {
                                 link_enter_failed(link);
-                                return;
+                                return r;
                         }
                         break;
+                case SD_DHCP_CLIENT_EVENT_SELECTING:
+                        if (dhcp_server_is_black_listed(link, client))
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: Then `!= 0`.

File: man/systemd.network.xml [hunk 21]
```diff
@@ -1566,6 +1566,13 @@
           </listitem>
         </varlistentry>
 
+        <varlistentry>
+          <term><varname>BlackList=</varname></term>
+          <listitem>
+            <para>A whitespace-separated list of ip addresses. Only accept packets from address[/cidr].</para>
```
Comment (yuwata) [inline man/systemd.network.xml]: `ip` -> `IPv4`

**Comment by ssahani** [comment]
yes updated thanks !

**Comment by yuwata** [comment]
Final minor updates are applied.

**Comment by yuwata** [comment]
bionic-amd64 failure is unrelated:
```
FAILED TESTS:  test/TEST-30-ONCLOCKCHANGE
autopkgtest [01:03:06]: test upstream: -----------------------]
autopkgtest [01:03:06]: test upstream:  - - - - - - - - - - results - - - - - - - - - -
upstream             FAIL non-zero exit status 1
```

**Pull request merged**